### PR TITLE
No Swap Rebalance strategies

### PR DESCRIPTION
**Pull request by dvpublic**
There are a lot of logs in contracts..

### CODE REVIEW

File: contracts/strategies/uniswap/UniswapV3ConverterStrategyLogicLib.sol [hunk 1]
```diff
@@ -856,4 +866,100 @@ library UniswapV3ConverterStrategyLogicLib {
...
+    address splitter
+  ) external returns (
+    uint[] memory tokenAmounts, // _depositorEnter(tokenAmounts) if length == 2
+    bool fuseEnabledOut
+  ) {
+    console.log("rebalanceNoSwaps.start");
+    uint loss;
+    tokenAmounts = new uint[](0);
+
+    if (state.fillUp) {
```
Comment (belbix) [inline contracts/strategies/uniswap/UniswapV3ConverterStrategyLogicLib.sol:884]: why?

File: contracts/strategies/uniswap/UniswapV3DebtLib.sol [hunk 2]
```diff
@@ -151,12 +166,64 @@ library UniswapV3DebtLib {
...
+
+    (p.newLowerTick, p.newUpperTick) = _calcNewTickRange(pool, state.lowerTick, state.upperTick, state.tickSpacing);
+    (p.prop0, p.prop1) = getEntryDataProportions(pool, p.newLowerTick, p.newUpperTick, p.depositorSwapTokens);
+
+    console.log("UniswapV3DebtLib.rebalanceNoSwaps.1");
+    BorrowLib.rebalanceAssets(
+      tetuConverter,
+      p.tokenA,
+      p.tokenB,
+      p.prop0 * 100_000 / (p.prop0 + p.prop1) // todo Probably rebalanceAssets should use original prop0, prop1
```
Comment (belbix) [inline contracts/strategies/uniswap/UniswapV3DebtLib.sol:191]: constant

Comment (belbix) [reply contracts/strategies/uniswap/UniswapV3DebtLib.sol:191]: what if p.prop0 + p.prop1 is zero?

File: contracts/strategies/uniswap/UniswapV3DebtLib.sol [hunk 3]
```diff
@@ -167,12 +234,9 @@ library UniswapV3DebtLib {
...
     // calculate proportions
     (uint consumed0, uint consumed1,) = UniswapV3Lib.addLiquidityPreview(address(pool), lowerTick, upperTick, token0Desired, token1Desired);
-
-    if (depositorSwapTokens) {
-      entryData = abi.encode(1, consumed1 * token1Price / token1Desired, consumed0);
-    } else {
-      entryData = abi.encode(1, consumed0, consumed1 * token1Price / token1Desired);
-    }
+    return depositorSwapTokens
+      ? (consumed1 * token1Price / token1Desired, consumed0)
```
Comment (belbix) [inline contracts/strategies/uniswap/UniswapV3DebtLib.sol:238]: token1Desired check on zero

File: contracts/libs/BorrowLib.sol [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    uint alpha;
+  }
+
+  /// @notice Set balances of {asset0} and {asset1} in proportions {proportion}:{100_000-proportion} using borrow/repay
+  /// @param proportion Proportion for {asset0}, [0...100_000]
+  function rebalanceAssets(
+    ITetuConverter tetuConverter_,
+    address asset0,
+    address asset1,
+    uint proportion
```
[the remark this answers is not in the document]
Comment (dvpublic) [reply contracts/libs/BorrowLib.sol:42]: I will replace 100_000 by 1e18

File: contracts/libs/BorrowLib.sol [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    console.log("openPosition.entryData.c.prop0", c.propA);
+    console.log("openPosition.entryData.c.prop1", c.propB);
+    console.log("openPosition.entryData.c.a02a1r", c.alpha);
+
+    if (balanceB_ != 0) {
+      // we are going to use {balanceA_} as collateral
+      // but there is some amount on {balanceB_}, so we need to keep corresponded part of {balanceA_} untouched
+      untouchedAmountA = (balanceB_ * c.alpha / 1e18) * c.propA / c.propB;
+      require(untouchedAmountA <= balanceA_, AppErrors.WRONG_VALUE);
```
[the remark this answers is not in the document]
Comment (dvpublic) [reply contracts/libs/BorrowLib.sol:177]: it's not possible, there is requirement: both props > 0, see require in rebalanceAssets()

**Pull request merged**